File: reference/android-source/java/dev/zanderp/opencfmoto/PxcHandshake.kt
package dev.zanderp.opencfmoto

import org.json.JSONObject
import java.net.Socket
import java.util.UUID

/**
 * Server-side PXC control dispatcher. The phone is the SERVER: after we send the
 * ECP_PXC_MDNS_RESPOND probe to the bike, the bike connects back to our listening
 * ports and drives the handshake. We reply per cmd.
 *
 * Verified against cfmoto-tcp-v5.log (official app, standard Car PXC, JSON CLIENT_INFO):
 *   bike 0x10000 (CAR_CTRL select)  -> we 0x10001
 *   bike 0x20000 (CAR_DATA select)  -> we 0x20001
 *   bike 0x10010 CLIENT_INFO (JSON) -> we 0x10011 (our info + RSA pubkey + signed HUID)
 *   bike 0x10690 {usbSpeed,wifiSpeed} -> we 0x10691
 *   bike 0x103e0 {client_set,sn}    -> we 0x103e1, then 0x201c0 CHECK_SN_RESULT {isOk:true}
 *   bike 0x70000000 heartbeat        -> we 0x70000001
 */
class PxcHandshake(
    private val log: (String) -> Unit,
) {
    private val phoneUuid: String = UUID.randomUUID().toString()
    @Volatile var carHuid: String? = null
        private set
    @Volatile var lastClientInfo: JSONObject? = null
        private set
    /** Dashboard-variant strategy, selected from the bike's CLIENT_INFO. Read by the media plane too
     *  (via EasyConnProber.handshake.profile). Written once on the control thread, before any media
     *  socket exists. */
    @Volatile var profile: BikeProfile = BikeProfiles.legacy
        private set
    /** Rate-limit HU_TIME_SYNC log lines (bike sends ~every 2s). */
    @Volatile private var lastHuTimeSyncLogAt: Long = 0L
    @Volatile private var huTimeSyncCount: Int = 0
    @Volatile private var clockLabBannerLogged: Boolean = false

    /**
     * Called when the bike selects a PXC channel on a :10922 socket (CAR_CTRL or CAR_DATA).
     * [EasyConnProber] starts a proactive 0x70000000 heartbeat on **each** of those sockets —
     * the 800NK (sdk 0.9.23.x) tears the whole session down after ~7s of silence on either one.
     */
    @Volatile var onPxcChannelSelected: ((Socket, String) -> Unit)? = null

    /** Dispatch one inbound frame on a given socket (ctrl or media). */
    fun handle(tag: String, frame: PxcFrame, socket: Socket) {
        val out = socket.getOutputStream()
        when (frame.cmd) {
            PxcFrame.CMD_CHANNEL_CAR_CTRL -> {
                log("[$tag] bike selected CAR_CTRL (0x10000) → ack 0x10001")
                PxcFrame(PxcFrame.CMD_CHANNEL_CAR_CTRL + 1, ByteArray(0)).write(out)
                onPxcChannelSelected?.invoke(socket, "CAR_CTRL")
            }
            PxcFrame.CMD_CHANNEL_CAR_DATA -> {
                log("[$tag] bike selected CAR_DATA (0x20000) → ack 0x20001")
                PxcFrame(PxcFrame.CMD_CHANNEL_CAR_DATA + 1, ByteArray(0)).write(out)
                onPxcChannelSelected?.invoke(socket, "CAR_DATA")
            }
            PxcFrame.CMD_CLIENT_INFO -> onClientInfo(tag, frame, out)
            PxcFrame.CMD_QUERY_SPEED -> {
                log("[$tag] QUERY_SPEED ${frame.payload.asText()} → reply 0x10691")
                PxcFrame(PxcFrame.CMD_QUERY_SPEED_RLY, ByteArray(0)).write(out)
            }
            PxcFrame.CMD_CHECK_SN -> onCheckSn(tag, frame, out)
            PxcFrame.CMD_HEARTBEAT -> {
                PxcFrame(PxcFrame.CMD_HEARTBEAT_ACK, ByteArray(0)).write(out)
            }
            PxcFrame.CMD_HEARTBEAT_ACK,
            PxcFrame.CMD_CHECK_SN_RESULT + 1 -> {
                // acks from the bike — nothing to do
            }
            // Clock lab: knobs choose 0x10600 echo vs phone and 0x10451
            // empty / Carbit / Zontes / no-ack. Defaults match Latest 2.0.13.
            PxcFrame.CMD_HU_TIME_SYNC -> onHuTimeSync(tag, frame, out)
            PxcFrame.CMD_HU_QUERY_TIME -> onHuQueryTime(tag, frame, out)
            else -> {
                if (!profile.handleUnknownControl(tag, frame, out, log)) {
                    log("[$tag] cmd=0x${frame.cmd.toUInt().toString(16)} (${PxcFrame.nameOf(frame.cmd)}) " +
                        "len=${frame.payload.size} ${frame.payload.asText()}")
                }
            }
        }
    }

    private fun logClockLabBanner() {
        if (clockLabBannerLogged) return
        clockLabBannerLogged = true
        val channel = lastClientInfo?.optString("channel")?.trim().orEmpty()
        log(ClockLab.banner(channel.ifEmpty { null }))
    }

    private fun onHuTimeSync(tag: String, frame: PxcFrame, out: java.io.OutputStream) {
        logClockLabBanner()
        val forcePhone = ClockLab.timeSync == ClockTimeSyncMode.PHONE
        val ack = HuTimeSync.ack(frame.payload, forcePhone = forcePhone)
        PxcFrame(PxcFrame.CMD_HU_TIME_SYNC_ACK, ack.payload).write(out)
        val n = ++huTimeSyncCount
        val now = System.currentTimeMillis()
        if (n <= 3 || now - lastHuTimeSyncLogAt >= 30_000L) {
            lastHuTimeSyncLogAt = now
            log("[CLOCK-LAB] HU_TIME_SYNC → ${ClockLab.timeSync.id}")
            log("[$tag] HU_TIME_SYNC #$n len=${frame.payload.size} → ack 0x10601 mode=${ack.mode} time=${ack.stamp}")
        }
    }

    private fun onHuQueryTime(tag: String, frame: PxcFrame, out: java.io.OutputStream) {
        logClockLabBanner()
        val channel = lastClientInfo?.optString("channel")?.trim().orEmpty()
        val mode = ClockLab.query
        val reply = when (mode) {
            ClockQueryMode.EMPTY -> "empty"
            ClockQueryMode.CARBIT, ClockQueryMode.ZONTES -> "json"
            ClockQueryMode.NO_ACK -> "no-ack"
        }
        log("[CLOCK-LAB] HU_QUERY_TIME len=${frame.payload.size} → 0x10451 $reply")
        when (mode) {
            ClockQueryMode.NO_ACK -> { /* tester: leave 0x10450 unanswered */ }
            ClockQueryMode.EMPTY -> {
                PxcFrame(PxcFrame.CMD_HU_QUERY_TIME_ACK, ByteArray(0)).write(out)
                log("[$tag] HU_QUERY_TIME (0x10450) len=${frame.payload.size} " +
                    "channel=${channel.ifEmpty { "-" }} → 0x10451 empty")
            }
            ClockQueryMode.CARBIT -> {
                val ack = HuQueryTime.carbit()
                PxcFrame(PxcFrame.CMD_HU_QUERY_TIME_ACK, ack.payload).write(out)
                log("[$tag] HU_QUERY_TIME (0x10450) len=${frame.payload.size} channel=${channel.ifEmpty { "-" }} " +
                    "→ 0x10451 carbit dateTime=${ack.dateTime}")
            }
            ClockQueryMode.ZONTES -> {
                val ack = HuQueryTime.zontesOem()
                PxcFrame(PxcFrame.CMD_HU_QUERY_TIME_ACK, ack.payload).write(out)
                log("[$tag] HU_QUERY_TIME (0x10450) len=${frame.payload.size} channel=${channel.ifEmpty { "-" }} " +
                    "→ 0x10451 zontes dateTime=${ack.dateTime} currentTime=${ack.currentTime} zone=${ack.timeZone}")
            }
        }
    }

    private fun onClientInfo(tag: String, frame: PxcFrame, out: java.io.OutputStream) {
        val text = frame.payload.asText()
        log("[$tag] *** CLIENT_INFO from bike *** $text")
        val json = try { JSONObject(text) } catch (e: Exception) {
            log("[$tag] CLIENT_INFO parse failed: $e"); return
        }
        lastClientInfo = json
        carHuid = json.optString("HUID").ifEmpty { json.optString("huid") }.ifEmpty { null }
        log("[$tag] carHuid=$carHuid HUName=${json.optString("HUName")} channel=${json.optString("channel")}")
        logClockLabBanner()

        profile = BikeProfiles.select(json, log)
        val early = BikeProfileHolder.active
        val startedSpec = BikeProfileHolder.aaVideo  // what Android Auto actually started / negotiated with
        // Keep a measured landscape panel (e.g. 800MT 1280×576) when CLIENT_INFO scoring would
        // flip to the near-square 800NK Advanced profile — that mis-route resized touch/margins
        // mid-session while AA stayed on the landscape stream (connect/drop flaps in field logs).
        val earlyPanel = early.panelSize
        val earlyIsWide = earlyPanel != null && earlyPanel.first >= earlyPanel.second * 3 / 2
        if (earlyIsWide && profile === Cfdl26NkTouchProfile) {
            log("[$tag] keeping landscape QR/measured profile '${early.name}' " +
                "(CLIENT_INFO preferred '${profile.name}' but panel ${earlyPanel!!.first}x${earlyPanel.second} is wide)")
            profile = early
        }
        if (early !== profile) {
            log("[$tag] profile refined from QR guess '${early.name}' → CLIENT_INFO '${profile.name}' " +
                "(AA already started at the QR-guess resolution)")
        }
        BikeProfileHolder.active = profile  // authoritative; QR modelId was only the early hint
        // Android Auto's video surface + decoder buffer were sized when AA started (from the QR guess)
        // and CANNOT be resized mid-session. If the refined profile wants a different resolution and no
        // explicit override is in effect, pin the spec to what AA is really running: otherwise the
        // compositor scales the live source (e.g. 800x480) as if it were the profile's (e.g. 720x1280),
        // producing a wildly wrong draw rect and a picture the dash rejects — which drops the link every
        // few seconds (the connect/drop flap). Known/consistent bikes hit no-op here.
        if (BikeProfileHolder.aaVideoOverride == null && BikeProfileHolder.aaVideo != startedSpec) {
            BikeProfileHolder.aaVideoOverride = startedSpec
            log("[$tag] pinned AA video to ${startedSpec.width}x${startedSpec.height} " +
                "(can't resize mid-session; refined profile wanted ${profile.aaVideo.width}x${profile.aaVideo.height})")
        }
        log("[$tag] *** BikeProfile selected: ${profile.name} ***")

        val reply = profile.buildClientInfoReply(json, carHuid, phoneUuid)
        log("[$tag] → CLIENT_INFO reply ${reply.toString().take(180)}…")
        PxcFrame(PxcFrame.CMD_CLIENT_INFO_RLY, reply.toString().toByteArray(Charsets.UTF_8)).write(out)
    }

    private fun onCheckSn(tag: String, frame: PxcFrame, out: java.io.OutputStream) {
        val text = frame.payload.asText()
        log("[$tag] CHECK_SN from bike: $text")
        val sn = try { JSONObject(text).optString("sn") } catch (e: Exception) { "" }
        // ack the request frame
        PxcFrame(PxcFrame.CMD_CHECK_SN_ACK, ByteArray(0)).write(out)
        // send the result
        val result = JSONObject().apply {
            put("isOk", true)
            put("errCode", 0)
            put("errMsg", "")
            put("id", sn)
            put("client_set", "easy_conn")
        }
        log("[$tag] → CHECK_SN_RESULT ${result}")
        PxcFrame(PxcFrame.CMD_CHECK_SN_RESULT, result.toString().toByteArray(Charsets.UTF_8)).write(out)
    }
}

private fun ByteArray.asText(): String =
    if (isEmpty()) "" else try { String(this, Charsets.UTF_8) } catch (e: Exception) { "<${size}b>" }

File: Sources/Protocol/PxcHandshake.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Part of AuraLink — an iOS port of OpenCfMoto (https://github.com/zanderp/open-cfmoto), AGPLv3.
// See LICENSE and NOTICE.
//
// Ported from PxcHandshake.kt. Server-side PXC control dispatcher: the phone is the SERVER — after
// EasyConnProber sends the ECP_PXC_MDNS_RESPOND probe to the bike, the bike connects back to our
// listening ports and drives the handshake; we only ever reply per-cmd. See
// docs/01-PROTOCOL-REFERENCE.md §3 for the full confirmed sequence.
//
// Verified against cfmoto-tcp-v5.log (Kotlin ground truth):
//   bike 0x10000 (CAR_CTRL select)   -> we 0x10001
//   bike 0x20000 (CAR_DATA select)   -> we 0x20001
//   bike 0x10010 CLIENT_INFO (JSON)  -> we 0x10011 (our info + RSA pubkey + signed HUID)
//   bike 0x10690 {usbSpeed,wifiSpeed}-> we 0x10691
//   bike 0x103e0 {client_set,sn}     -> we 0x103e1, then we proactively send 0x201c0 {isOk:true}
//   bike 0x70000000 heartbeat        -> we 0x70000001
//
// An `actor` because CAR_CTRL and CAR_DATA arrive on two independent sockets/Tasks but both mutate
// shared state (`profile`, `carHuid`, `lastClientInfo`) — the actor serializes that for free instead
// of hand-rolling a lock, matching the project's "async/await, not manual thread-safety" convention.

import Foundation

actor PxcHandshake {

    /// What the caller (EasyConnProber) should do after `handle(...)` returns.
    struct Outcome: Sendable {
        /// Frames to write back to the SAME socket the inbound frame arrived on, in order.
        let replies: [PxcFrame]
        /// Set when this frame was a channel-select (CAR_CTRL/CAR_DATA) — the caller should start a
        /// proactive 2s heartbeat sender on this socket. See docs/01-PROTOCOL-REFERENCE.md §3 step 5:
        /// leaving EITHER channel socket unheartbeated caused a real 800NK-family ~7s disconnect flap
        /// in the Android app, so heartbeat both, unconditionally, not just on CAR_CTRL.
        let channelSelected: String?

        static func reply(_ frames: PxcFrame...) -> Outcome { Outcome(replies: frames, channelSelected: nil) }
        static func channel(_ frame: PxcFrame, name: String) -> Outcome { Outcome(replies: [frame], channelSelected: name) }
        static let none = Outcome(replies: [], channelSelected: nil)
    }

    private let phoneUUID: String
    private(set) var carHuid: String?
    private(set) var lastClientInfo: BikeClientInfo?
    private(set) var profile: BikeProfile = BikeProfiles.default

    init(phoneUUID: String = PxcHandshake.persistentPhoneUUID()) {
        self.phoneUUID = phoneUUID
    }

    /// Dispatch one inbound control-plane frame. `tag` is a short label for log lines (e.g. ":10922").
    func handle(tag: String, frame: PxcFrame) async -> Outcome {
        switch frame.cmd {
        case PxcCmd.channelCarCtrl:
            await LogBus.shared.log("[\(tag)] bike selected CAR_CTRL (0x10000) → ack 0x10001")
            return .channel(PxcFrame(cmd: PxcCmd.channelCarCtrl &+ 1), name: "CAR_CTRL")

        case PxcCmd.channelCarData:
            await LogBus.shared.log("[\(tag)] bike selected CAR_DATA (0x20000) → ack 0x20001")
            return .channel(PxcFrame(cmd: PxcCmd.channelCarData &+ 1), name: "CAR_DATA")

        case PxcCmd.clientInfo:
            return await onClientInfo(tag: tag, frame: frame)

        case PxcCmd.querySpeed:
            await LogBus.shared.log("[\(tag)] QUERY_SPEED \(frame.payloadText) → reply 0x10691")
            return .reply(PxcFrame(cmd: PxcCmd.querySpeedReply))

        case PxcCmd.checkSn:
            return await onCheckSn(tag: tag, frame: frame)

        case PxcCmd.heartbeat:
            return .reply(PxcFrame(cmd: PxcCmd.heartbeatAck))

        case PxcCmd.heartbeatAck, PxcCmd.checkSnResultAck:
            return .none // acks from the bike — nothing to do

        case PxcCmd.huTimeSync:
            return await onHuTimeSync(tag: tag, frame: frame)

        case PxcCmd.huQueryTime:
            await LogBus.shared.log("[\(tag)] HU_QUERY_TIME (0x10450) len=\(frame.payload.count) → 0x10451 empty")
            return .reply(PxcFrame(cmd: PxcCmd.huQueryTimeAck, payload: HuQueryTime.emptyAck()))

        default:
            if let reply = profile.handleUnknownControl(cmd: frame.cmd, payload: frame.payload) {
                await LogBus.shared.log(
                    "[\(tag)] cmd=\(frame.cmdHex) (\(PxcCmd.name(of: frame.cmd))) len=\(frame.payload.count) "
                        + "\(frame.payloadText) → ack \(reply.cmdHex)"
                )
                return .reply(reply)
            }
            await LogBus.shared.log(
                "[\(tag)] cmd=\(frame.cmdHex) (\(PxcCmd.name(of: frame.cmd))) len=\(frame.payload.count) \(frame.payloadText)"
            )
            return .none
        }
    }

    private func onClientInfo(tag: String, frame: PxcFrame) async -> Outcome {
        let text = frame.payloadText
        await LogBus.shared.log("[\(tag)] *** CLIENT_INFO from bike *** \(text)")
        guard let info = BikeClientInfo.parse(frame.payload) else {
            await LogBus.shared.log("[\(tag)] CLIENT_INFO parse failed — no valid JSON in payload")
            return .none
        }
        lastClientInfo = info
        carHuid = info.huid
        await LogBus.shared.log("[\(tag)] carHuid=\(carHuid ?? "nil") HUName=\(info.huName) channel=\(info.channel) sdkVersion=\(info.sdkVersion)")

        let (selected, scoreLog) = BikeProfiles.select(info)
        await LogBus.shared.log(scoreLog)
        profile = selected
        await LogBus.shared.log("[\(tag)] *** BikeProfile selected: \(selected.name) ***")

        let reply = PhoneClientInfo.build(
            huid: carHuid,
            phoneUUID: phoneUUID,
            supportFunction: selected.advertisedSupportFunction,
            advertiseTouch: selected.supportsScreenTouch
        )
        let json: Data
        do {
            json = try reply.jsonData()
        } catch {
            await LogBus.shared.log("[\(tag)] CLIENT_INFO reply encode failed: \(error) — sending empty reply")
            json = Data()
        }
        let preview = String(data: json, encoding: .utf8).map { String($0.prefix(180)) } ?? ""
        await LogBus.shared.log("[\(tag)] → CLIENT_INFO reply \(preview)…")
        return .reply(PxcFrame(cmd: PxcCmd.clientInfoReply, payload: json))
    }

    private func onCheckSn(tag: String, frame: PxcFrame) async -> Outcome {
        let text = frame.payloadText
        await LogBus.shared.log("[\(tag)] CHECK_SN from bike: \(text)")
        let sn: String
        if let obj = try? JSONSerialization.jsonObject(with: frame.payload) as? [String: Any] {
            sn = (obj["sn"] as? String) ?? ""
        } else {
            sn = ""
        }
        let ack = PxcFrame(cmd: PxcCmd.checkSnAck)
        let resultObj: [String: Any] = ["isOk": true, "errCode": 0, "errMsg": "", "id": sn, "client_set": "easy_conn"]
        let resultData = (try? JSONSerialization.data(withJSONObject: resultObj)) ?? Data()
        await LogBus.shared.log("[\(tag)] → CHECK_SN_RESULT \(String(data: resultData, encoding: .utf8) ?? "")")
        return Outcome(replies: [ack, PxcFrame(cmd: PxcCmd.checkSnResult, payload: resultData)], channelSelected: nil)
    }

    private func onHuTimeSync(tag: String, frame: PxcFrame) async -> Outcome {
        let ack = HuTimeSync.ack(for: frame.payload)
        await LogBus.shared.log("[\(tag)] HU_TIME_SYNC len=\(frame.payload.count) → ack 0x10601 mode=\(ack.mode) time=\(ack.stamp)")
        return .reply(PxcFrame(cmd: PxcCmd.huTimeSyncAck, payload: ack.payload))
    }

    /// Per-install phone UUID, generated once and persisted — the Kotlin app generates a fresh
    /// `UUID.randomUUID()` per PROCESS (not persisted); we persist instead since the protocol field
    /// is explicitly documented as "generate+persist per install" in docs/01-PROTOCOL-REFERENCE.md §3.
    /// A full `AppSettings` (Phase 6) will likely own this key instead — left here as a self-contained
    /// default so PxcHandshake has no dependency on a not-yet-written settings module.
    static func persistentPhoneUUID() -> String {
        let key = "com.amielsena.auralink.phoneUUID"
        if let existing = UserDefaults.standard.string(forKey: key) { return existing }
        let fresh = UUID().uuidString
        UserDefaults.standard.set(fresh, forKey: key)
        return fresh
    }
}
